controllers/azurejson_machine_controller.go

```diff
@@ -99,14 +99,18 @@ func (f filterUnclonedMachinesPredicate) Generic(e event.GenericEvent) bool {
 func (r *AzureJSONMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
 	defer cancel()
-	log := r.Log.WithValues("namespace", req.Namespace, "azureMachine", req.Name)
 
-	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureJSONMachineReconciler.Reconcile",
+	ctx, span, log := tele.StartSpanWithLogger(
+		ctx,
+		"controllers.AzureJSONMachineReconciler.Reconcile",
+		tele.Tracer(),
```

Comment:
I think I understand why you are adding `tele.Tracer()` and `r.Log` to the func signature. The signature you have exposed here makes it clear what the func depends upon. Transparency and ease of testing are always good.

The thing I struggle with is that if this function is used in every place with the same input, then perhaps, we need two functions. One that has all of these inputs, and another that defaults some of the inputs.

Further, the `r.Log` field should actually go away. We should not initialize a logger field on the reconciler, but rather use the context to carry the logger. I can see how this would be opaque to a caller, but it would be more consistent with [Cluster API](https://github.com/kubernetes-sigs/cluster-api/blob/074c2cb334ab54e4f8ef5ddcf40a15a3aa1044a3/controllers/cluster_controller.go#L97). Being part of the Cluster API ecosystem and staying as consistent as we can with CAPI would be a kindness to developers contributing across the projects.

Perhaps, this fits with having two functions, one that defaults in an opinionated way, and another that is referentially transparent. 

```go
      ctx, span, logger := tele.StartSpanWithLogger(
		ctx,
		"controllers.AzureMachineReconciler.Reconcile",
		trace.WithAttributes(
			attribute.String("namespace", req.Namespace),
			attribute.String("name", req.Name),
			attribute.String("kind", "AzureMachine"),
		),
	)
```

Perhaps, we could even go a step further. Since we only use one func on the `Span`, and that it would beneficial to get logging statements on the span (as you have done with the composite logger below :+1: ), consider returning a `SpanLogger`. Similar to the following.

```go
      ctx, spanLogger := tele.StartSpanWithLogger(
		ctx,
		"controllers.AzureMachineReconciler.Reconcile",
		trace.WithAttributes(
			attribute.String("namespace", req.Namespace),
			attribute.String("name", req.Name),
			attribute.String("kind", "AzureMachine"),
		),
	)
        defer spanLogger.End()
        spanLogger.Infof("something to log on both span and logger named %q", thing)
```

@CecileRobertMichon wdyt?

Reply:
+1 for having a default alternate. Not sure about the 2nd point though as we'll need the `compositeLogger` to embed `trace.Span`, which is not the intent here iiuc.

Reply:
> The thing I struggle with is that if this function is used in every place with the same input, then perhaps, we need two functions. One that has all of these inputs, and another that defaults some of the inputs.

+1, seems like a good improvement

Not sure I follow the rest, @shysank can you help me understand what you mean by 
> we'll need the compositeLogger to embed trace.Span, which is not the intent here iiuc.

?

Reply:
Maybe I misunderstood David's comment. My understanding was that the logger `spanLogger` returned from `StartSpanWithLogger` will still be a composite logger of some sort with `trace.Span` in it to call `span.end()`. So, either we modify `compositeLogger` to include `trace.Span`, or modify spanLogger to include composite logger functionality, both of which seem to be coupling two different things imo.

Reply:
> One that has all of these inputs, and another that defaults some of the inputs.

We can avoid adding a second function to this API with functional options or similar. Would that be sufficient?

>Further, the r.Log field should actually go away. We should not initialize a logger field on the reconciler, but rather use the context to carry the logger

What do you mean by "initialize a logger"? `StartSpanWithLogger` creates two sub-loggers from the incoming `r.Log`.

The overall problem I have with this approach is the implicitness, as you probably could have guessed. We trade writing one extra parameter per call site -- more work to write and review initially, to be sure -- for a holistic increase in readability. Also, context values are intended for request scoped data, not for parameters to functions.

One other note - I don't see what is gained by using the embedded logger, from an API or community standpoint. This function wouldn't be any more compatible with the cluster API ecosystem if it simply extracted the logger from the context, would it?

>consider returning a `SpanLogger`

the `SpanLogger` as is currently written, or something different? as currently written, the `SpanLogger` won't write to standard out, which seems problematic.

>modify spanLogger to include composite logger functionality, both of which seem to be coupling two different things imo

agreed with you @shysank - my intended purpose for `spanLogger` is logging, not tracing. I think there would be a lot of long-term benefit to keeping those two functional areas orthogonal. The same applies to span logging and logs to standard out as well.

Reply:
The composite logger already has the span embedded. Sure, the logger being returned fits the interface for a logger, but it has a dependency on both, std logger and span logger. They are already connected explicitly. Each log message that goes to standard out also goes to the span. The only thing missing is the `End()` func. 

As for fetching the logger from the context, it would be much better than the options we have now. At this point, we add the logger as field on structs, so we can avoid passing the logger into all of the methods all the way down the chain of calls. This is precisely where context works for us, because it is the one thing we have chosen to pass all the way down the call chain.

Reply:
>The composite logger already has the span embedded

Yup, my intention with the composite logger was to provide it as a union of N loggers. The span and standard loggers are just `logr.Logger` implementations to the composite logger, and I think this setup probably shouldn't know anything about tracing.

>it would be much better than the options we have now

regarding this and the above point, we might be able to come to a good compromise by:

- Abstracting the magic that I don't like, while also...
- Providing an API that uses it under the hood and...
- Giving users the ability to easily `defer span.End()` while not tightly coupling logging and tracing concerns
Here's an idea:

```go
func StartSpan(
	ctx context.Context,
	spanName string,
	opts ...trace.SpanOption,
) (context.Context, logr.Logger, func()) {
	lggr := extractLoggerFromCtx(ctx) // this abstracts the "magic" of pulling a logger out of a context.Context
	c, span := tele.Tracer().Start(ctx, spanName, opts...)
	endFn := func() {
		span.End()
	}
	return c, &compositeLogger{
		loggers: []logr.Logger{
			corrIDLogger(ctx, lggr),
			&spanLogger{Span: span},
		},
	}, endFn
}
```

In reconcilers, this would be called as:

```go
ctx, log, done := StartSpan(ctx, "MyReconciler")
defer done()

log.Info("Starting!")
```

This follows a similar pattern as `context.WithCancel`, `context.WithTimeout`, etc... with its "done" return function, while hiding the existence of a Span completely. Users simply need to defer a call to the final return value and nothing else. It also abstracts and centralizes the functionality that pulls loggers out of the context.

WDYT?

Reply:
Great suggestion! 👍

Reply:
What do other think of https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1575#discussion_r704841410?

Reply:
I like it, the done() func is really neat

Reply:
+1 from me as well.

Reply:
@devigned @CecileRobertMichon @shysank I'll build this out in a distinct commit within this PR, so that people can get a feel for the ergonomics

Reply:
ok, see https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1575/commits/26c88a13d71d2c65264d6bbd44a3ae806c0f55b9